multiview/embed/mvmds.py

```diff
@@ -0,0 +1,217 @@
+    eigenvectors are returned based on the algorithm proposed by the
+    following paper:
+
+    https://www.sciencedirect.com/science/article/pii/S016794731000112X
+
+    """
+
+    def __init__(self, n_components=None):
+
+        super().__init__()
+        self.components = None
+        self.n_components = n_components
+
+    def _cpc(self, x):
+
+        """
+        Finds Stepwise Estimation of Common Principal Components as described
+        by common Trendafilov implementations based on the following paper:
+
+        https://www.sciencedirect.com/science/article/pii/S016794731000112X
+
+        Parameters
+        ----------
+
+        x: List of matrices, each with number of rows, n
+
+        it: Number of common principal component stepwise iterations
+
+        Returns
+        -------
+
+        Components: Desired Common Principal Components
+
+        """
+        n = p = x.shape[1]
+
+        views = len(x)
+
+        n_num = np.array([n] * views)/np.sum(np.array([n] * views))
+
+        Components = np.zeros((p, self.n_components))
+
+        pi = np.eye(p)
+
+        s = np.zeros((p, p))
+
+        it = 15
+
+        for i in np.arange(views):
+            s = s + (n_num[i] * x[i])
+
+        e1, e2 = np.linalg.eigh(s)
+
+        q0 = e2[:, ::-1]
+
+        for i in np.arange(self.n_components):
+
+            q = q0[:, i]
+            q = np.array(q).reshape(len(q), 1)
+            d = np.zeros((1, views))
+
+            for j in np.arange(views):
+
+                d[:, j] = np.dot(np.dot(q.T, x[j]), q)
+
+            for j in np.arange(it):
+                s2 = np.zeros((p, p))
+
+                for yy in np.arange(views):
+                    d2 = n_num[yy] * np.sum(np.array([n] * views))
+
+                    if d[:, yy] == 0:
+                        s2 = s2 + (d2 * x[yy] / .0001)
+
+                    else:
+                        s2 = s2 + (d2 * x[yy] / d[:, yy])
+
+                w = np.dot(s2, q)
+
+                w = np.dot(pi, w)
+
+                q = w / np.sqrt(np.dot(w.T, w))
+
+                for yy in np.arange(views):
+
+                    d[:, yy] = np.dot(np.dot(q.T, x[yy]), q)
+
+            Components[:, i] = q[:, 0]
+            pi = pi - np.dot(q, q.T)
+
+        return(Components)
+
+    def fit(self, Xs):
+
+        """
+        Calculates a dimensional reduction based on minimizing Euclidean
+        Distances of multiple views
+
+        Parameters
+        ----------
+
+        Xs: list of array-likes
+                - Xs shape: (n_views,)
+                - xXs[i] shape: (n_samples, n_features_i)
+                The data to fit to. Each sample will receive its own embedding.
+
+        n_components: int (positive), optional, default = 0
+            Number of components requested
+
+        Attributes
+        ----------
+
+        components: A k-dimensional projection of shape [n,k]
+
+        """
+
+        if (self.n_components) > len(Xs[0]):
```

Comment:
For the future, all these checks should be put into a private function and then called here. However, for the sake of the time I think it's fine as is. Maybe add an issue for now

Reply:
+ 1

Reply:
Will add as an issue